File: packages/js-dpp/lib/stateTransition/fees/operations/WriteOperation.js
const AbstractOperation = require('./AbstractOperation');

class WriteOperation extends AbstractOperation {
  /**
   * @param {number} keySize
   * @param {number} valueSize
   */
  constructor(keySize, valueSize) {
    super();

    this.keySize = keySize;
    this.valueSize = valueSize;
  }

  /**
   * Get CPU cost of the operation
   *
   * @returns {number}
   */
  getProcessingCost() {
    return ((this.keySize + this.valueSize) * AbstractOperation.STORAGE_PROCESSING_CREDIT_PER_BYTE)
      + WriteOperation.BASE_PROCESSING_COST;
  }

  /**
   * Get storage cost of the operation
   *
   * @returns {number}
   */
  getStorageCost() {
    return (this.keySize + this.valueSize) * AbstractOperation.STORAGE_CREDIT_PER_BYTE;
  }

  /**
   * Get operation type
   *
   * @returns {string}
   */
  getType() {
    return AbstractOperation.TYPES.WRITE;
  }
}

WriteOperation.BASE_PROCESSING_COST = 60000;

module.exports = WriteOperation;

File: packages/rs-dpp/src/state_transition/fees/operations/write_operation.rs
use super::{
    OperationLike, OperationType, STORAGE_CREDIT_PER_BYTE, STORAGE_PROCESSING_CREDIT_PER_BYTE,
};

pub const BASE_PROCESSING_COST: i64 = 60000;

#[derive(Default, Debug, Clone, Copy)]
pub struct WriteOperation {
    key_size: usize,
    value_size: usize,
}

impl WriteOperation {
    pub fn new(key_size: usize, value_size: usize) -> Self {
        Self {
            key_size,
            value_size,
        }
    }
}

impl OperationLike for WriteOperation {
    fn get_processing_cost(&self) -> i64 {
        BASE_PROCESSING_COST
            + ((self.key_size + self.value_size) as i64 * STORAGE_PROCESSING_CREDIT_PER_BYTE)
    }

    fn get_storage_cost(&self) -> i64 {
        (self.key_size + self.value_size) as i64 * STORAGE_CREDIT_PER_BYTE
    }

    fn get_type(&self) -> OperationType {
        OperationType::Write
    }
}
